kernel/src/device/tdxguest/mod.rs

```diff
@@ -176,36 +221,40 @@ fn handle_get_report(arg: usize) -> Result<i32> {
     let report = tdx_get_report(&user_request.report_data)?;
 
     let tdx_report_vaddr = arg + TDX_REPORTDATA_LEN;
-    user_space.write_bytes(tdx_report_vaddr, &mut VmReader::from(report.as_ref()))?;
+    user_space.write_bytes(tdx_report_vaddr, report.reader().limit(TDX_REPORT_LEN))?;
     Ok(0)
 }
 
-fn tdx_get_report(inblob: &[u8]) -> Result<Box<[u8]>> {
+/// Gets the TDX report given the specified data in `inblob`.
+///
+/// The first `TDX_REPORT_LEN` bytes of data in the returned `USegment` is the report.
+/// The rest in `USegment` should be ignored.
+fn tdx_get_report(inblob: &[u8]) -> Result<USegment> {
     if inblob.len() != TDX_REPORTDATA_LEN {
         return_errno_with_message!(Errno::EINVAL, "Invalid inblob length");
     }
 
-    let segment = FrameAllocOptions::new().alloc_segment(2)?;
-    let dma_coherent = DmaCoherent::map(segment.into(), false).unwrap();
-    dma_coherent.write_bytes(0, &inblob).unwrap();
+    let report: USegment = FrameAllocOptions::new().alloc_segment(1)?.into();
```

Comment:
When I read the code for the first time, it got me wonder whether one page is large enough. The following code would make this point more clear.

```rust
let report: USegment = {
    const REPORT_SIZE_IN_PAGES: usize = size_of::<TdReport>().div_ceil(PAGE_SIZE);
    FrameAllocOptions::new().alloc_segment(REPORT_SIZE_IN_PAGES)?.into()
};
```

Note that the above code is as efficient as the current one because `REPORT_NR_PAGES` can be calculated at the runtime.